### ISSUE TITLE
Increasing buffer

### ISSUE THREAD

**Issue by thegobot** [begeekmyfriend/yasea#85]
Hello!

I have a problem with increasing the buffer in your player
I use the following settings:
bufferTime = 0.5
bufferTimeMax = 3

1) Player is connected(adobe flash), and stream has not yet been published
2) Stream was published
3) Player received NetStream.Play.PublishNotifi event and began to play the stream. Buffer ~3...5...60 sec. The greater the restart stream, the more increases the buffer

When the player is connected to the already published stream - OK. Buffer ~ 0.5-1 sec.

The speed of the server to the client is stable, I measured (~ 595 KB)

Also, I have my own streamer written in adobe flash. This player does not have this problem with this buffer. He always stable ~ 0.5-1 sec. and it does not grow when restarts the stream

also observe a message
`W/RtmpConnection: handleRxPacketLoop(): Not handling unimplemented/unknown packet of type: ACKNOWLEDGEMENT`

**Comment by begeekmyfriend** [comment]
I do nothing with player.

`ACKNOWLEDGEMENT` does not need to be handled.

**Comment by thegobot** [comment]
I think I found the reason
startEncode() , startAudio(),  addCallbackBuffer(), setPreviewCallbackWithBuffer() does not need to run after the initialization of the camera, and at the time of publication stream.
Also, after sop stream need stopEncode(), stopAudio() and setPreviewCallback(null)

Also, the problem is solved on the player's side. you need to play with a delay of 3-5 sec., after PublishNotify

Abstract code Avtion Script 3

NetStream.Play.UnpublishNotify:
stream.pause()

NetStream.Play.PublishNotify:
setTimeout(function() : void{
    stream.resume();
}, 3333);

otherwise, after publish stream the dropped frames will grow and increase buffer

**Comment by thegobot** [comment]
The time between the start of the encoder and NetStream.Publish.Start this time to increase the buffer. If I run the encoder after NetStream.Publish.Start that the player does not increase the buffer. I do not know much about the protocol can be the case in calculating a PTS or DTS?

I artificially made on the server and delay the publication time of this delay is strictly corresponds to an increase of the buffer in the player.

The problem exists and it is very important when the stream resumes...

**Comment by begeekmyfriend** [comment]
The first video frame after RTMP connection must be IDR frame, otherwise the display will be flower screen. I recommend you to testify it on [bbs.chinaffmpeg.com](http://bbs.chinaffmpeg.com/1.swf)

**Comment by thegobot** [comment]
This player I have the same delay
It's simple. If there is a delay(wait > 500ms... 3-5sec) in createStream https://github.com/begeekmyfriend/yasea/blob/master/app/src/main/java/net/ossrs/yasea/rtmp/io/RtmpConnection.java#L220 is large, the buffer player grow just enough

The delay in publication of the flow on the production server can be up to 5 seconds, and it must be provided

**Comment by thegobot** [comment]
1) connect(rtmpUrl) before loop https://github.com/thegobot/yasea/blob/master/app/src/main/java/net/ossrs/yasea/SrsFlvMuxer.java#L157
2) startEncoder() && startAudio() after NetStream.Publish.Start, see https://github.com/begeekmyfriend/yasea/pull/112/commits/0cebc142a4110dc95a0dec156148db5223805142

Then there is no delay in the player (buffer ~ 1s - is ok). Remember, increasing the buffer appears only after the resume of the Stream. If you start playback when the stream was already running - no such problem

**Comment by begeekmyfriend** [comment]
Definitely. Please try this [commit](https://github.com/begeekmyfriend/yasea/commit/9339e9d5442281c84a5882a831ba675637348fcd). Thank you very much;-)

**Comment by thegobot** [comment]
Thanks!
For me it did not solve the problem ... I will try to explain again ... sorry for the English
The scheme is
Click on the Start button
1) Start encoder, frames went immediately into the cache(frameCache).
2) Connect to rtmp server...~0.1sec... 
3) Publish the stream and wait NetStream.Publish.Start... 3-5 sec(!!!)...
4) During this time, the cache continues to accumulate
5) handleRxInvoke received a reply NetStream.Publish.Start - start the loop and sendFlvTag()

Pause between the connections and sending flv frames increase the buffer in the player

A solution that works:
Adding frames to the cache AFTER NetStream.Publish.Start. They should not be added until the server is received NetStream.Publish.Start. On the production servers compound may be 3 seconds or more.

Also in the player after receiving NetStream.Play.PublishNotify timeline is set to 5 seconds. Therefore, the client thinks that with communication problems and increases the buffer (and then it is reduced to the normal range)

Try the following:
1) Start broadcast
1) Launch Player. Ок, buffer ~1sec
2) Stop broadcast
3) Start broadcast. In this step, the player will receive NetStream.Play.PublishNotifi event and resume playback, In this step, the player will receive NetStream.Play.PublishNotifi event and resume playback, but will increase the buffer that will result in the flow of delay

**Comment by begeekmyfriend** [comment]
···java
Connect to rtmp server
09-06 01:07:27.848 WriteThread: getTimestampDelta: 0, AbsoluteTimestamp: 5
09-06 01:07:28.248 WriteThread: getTimestampDelta: -1, AbsoluteTimestamp: 417
09-06 01:07:28.258 WriteThread: getTimestampDelta: -1, AbsoluteTimestamp: 422
09-06 01:07:28.258 WriteThread: getTimestampDelta: -1, AbsoluteTimestamp: 422
09-06 01:07:28.258 WriteThread: getTimestampDelta: -1, AbsoluteTimestamp: 423
09-06 01:07:28.538 WriteThread: getTimestampDelta: -1, AbsoluteTimestamp: 701
WAIT NetStream.Publish.Start
···

These RTMP packets belong to commands but not AV streams so that it should not be count into delay. Also there is no extra packets being injected into the buffer of the player during waiting for the `NetStream.Publish.Start` response. **8 seconds delay on player is absolutely non-sense.**

Nevertheless, I still improve the process of connection and packet sending so that you can try it again https://github.com/begeekmyfriend/yasea/commit/b1f35ffe89875279b91faa9c19d1ae6074457dbc

**Comment by thegobot** [comment]
The fact that the buffer begins to accumulating frames before starting network operation. In my example, the frames begin to be transferred after ~ 5 seconds. For 5 seconds buffer time to fill up the data, but they are irrelevant. Why send them? Starting the encoder must be AFTER NetStream.Publish.Start. Then be transferred to the actual data, without delay.

**Comment by begeekmyfriend** [comment]
The AV packet will not be sent [until](https://github.com/begeekmyfriend/yasea/blob/master/app/src/main/java/net/ossrs/yasea/SrsFlvMuxer.java#L123) the `publish` method [returns](https://github.com/begeekmyfriend/yasea/blob/master/app/src/main/java/net/ossrs/yasea/rtmp/io/RtmpConnection.java#L244).

**Comment by thegobot** [comment]
```
 I/ACodec: setupVideoEncoder succeeded
 D/RtmpConnection: connect(): socket connection established, doing handhake...
 D/Handshake: writeC0
 D/Handshake: writeC1
 D/Handshake: writeC1(): Calculating digest offset
 D/Handshake: writeC1(): (real value of) digestOffset: 414
 D/Handshake: writeC1(): recalculated digestOffset: 1190
 D/Handshake: writeC1(): digestOffsetBytes: 00009FFF
 D/Handshake: partBeforeDigest(): size: 1190
 D/Handshake: writeC1(): Writing timestamp and Flash Player version
 D/Handshake: partAfterDigest(): size: 314
 D/Handshake: copying digest offset bytes in partBeforeDigest
 D/Handshake: writeC1(): Calculating digest
 D/Handshake: writeC1(): writing C1 packet
 D/Handshake: readS0
 D/Handshake: readS1
 D/Handshake: readS1(): S1 total bytes read OK
 D/Handshake: readC2
 D/Handshake: readS2
 ```
**E/SrsPublisher: ENCODER start** - wrong
 ```
 D/Handshake: readS2(): S2 total bytes read OK
 D/RtmpConnection: connect(): handshake done
 D/RtmpConnection: rtmpConnect(): Building 'connect' invoke packet
 D/RtmpConnection: starting main rx handler loop
 D/RtmpConnection: wrote packet: RTMP Command (command: connect, transaction ID: 1), size: 257
 D/RtmpDecoder: readPacket(): header.messageType: WINDOW_ACKNOWLEDGEMENT_SIZE
 D/RtmpConnection: handleRxPacketLoop(): Setting acknowledgement window size: 5000000
```
**Encoder is already running, adding frames to the cache:**
I/SrsFlvMuxer: annexb demux 25B, pts=0, frame=13B, nalu=7
 I/SrsFlvMuxer: annexb demux 25B, pts=0, frame=4B, nalu=8
 I/SrsFlvMuxer: flv: h264 sps/pps sent, sps=13B, pps=4B
```
 D/RtmpDecoder: readPacket(): header.messageType: SET_PEER_BANDWIDTH
 D/RtmpConnection: handleRxPacketLoop(): Send acknowledgement window size: 5000000
 D/RtmpDecoder: readPacket(): header.messageType: SET_CHUNK_SIZE
 D/RtmpConnection: wrote packet: RTMP Window Acknowledgment Size, size: 4
 D/RtmpDecoder: readPacket(): Setting chunk size to: 4096
 D/RtmpDecoder: readPacket(): header.messageType: COMMAND_AMF0
 D/RtmpConnection: handleRxInvoke: Got result for invoked method: connect
 D/RtmpConnection: createStream(): Sending releaseStream command...
 D/RtmpConnection: wrote packet: RTMP Command (command: releaseStream, transaction ID: 2), size: 92
 D/RtmpConnection: createStream(): Sending FCPublish command...
 D/RtmpConnection: wrote packet: RTMP Command (command: FCPublish, transaction ID: 3), size: 88
 D/RtmpConnection: createStream(): Sending createStream command...
```
Pause ~5 sec.
```
 D/RtmpConnection: wrote packet: RTMP Command (command: createStream, transaction ID: 4), size: 25
```
NetStream.Publish.Start сommand is received  // We can now publish AV data
# Here it is necessary to run the encoder
```
 D/RtmpDecoder: readPacket(): header.messageType: COMMAND_AMF0
 D/RtmpConnection: handleRxInvoke: Got result for invoked method: createStream
 D/RtmpConnection: handleRxInvoke(): Stream ID to publish: 1
 D/RtmpConnection: fmlePublish(): Sending publish command...
 D/RtmpConnection: wrote packet: RTMP Command (command: publish, transaction ID: 0), size: 93
 D/RtmpDecoder: readPacket(): header.messageType: COMMAND_AMF0
 D/RtmpConnection: onMetaData(): Sending empty onMetaData...
 D/RtmpConnection: wrote packet: net.ossrs.yasea.rtmp.packets.Data@4257c3d0, size: 238
 D/RtmpConnection: wrote packet: RTMP Video, size: 33
 I/SrsFlvMuxer: worker: send frame type=9, dts=0, size=33B
 D/RtmpConnection: wrote packet: RTMP Video, size: 5
 D/RtmpConnection: wrote packet: RTMP Audio, size: 127
 D/RtmpConnection: wrote packet: RTMP Video, size: 5150
 I/SrsFlvMuxer: worker: send frame type=9, dts=609, size=5150B
 D/RtmpConnection: wrote packet: RTMP Audio, size: 126
 D/RtmpConnection: wrote packet: RTMP Video, size: 5679
 D/RtmpConnection: wrote packet: RTMP Audio, size: 124
 D/RtmpConnection: wrote packet: RTMP Audio, size: 129
 D/RtmpConnection: wrote packet: RTMP Audio, size: 128
 D/RtmpConnection: wrote packet: RTMP Audio, size: 121
 D/RtmpConnection: wrote packet: RTMP Audio, size: 120
 D/RtmpConnection: wrote packet: RTMP Audio, size: 115
 D/RtmpConnection: wrote packet: RTMP Video, size: 4406
 D/RtmpConnection: wrote packet: RTMP Video, size: 809
 D/RtmpConnection: wrote packet: RTMP Audio, size: 113
 D/RtmpConnection: wrote packet: RTMP Audio, size: 115
 D/RtmpConnection: wrote packet: RTMP Audio, size: 98
 D/RtmpConnection: wrote packet: RTMP Audio, size: 108
 D/RtmpConnection: wrote packet: RTMP Audio, size: 98
 D/RtmpConnection: wrote packet: RTMP Video, size: 3562
 D/RtmpConnection: wrote packet: RTMP Audio, size: 104
 D/RtmpConnection: wrote packet: RTMP Video, size: 5058
 D/RtmpConnection: wrote packet: RTMP Audio, size: 99
 D/RtmpConnection: wrote packet: RTMP Audio, size: 100
 D/RtmpConnection: wrote packet: RTMP Audio, size: 122
 D/RtmpConnection: wrote packet: RTMP Audio, size: 116
 D/RtmpConnection: wrote packet: RTMP Video, size: 3733
 D/RtmpConnection: wrote packet: RTMP Video, size: 2724
 D/RtmpConnection: wrote packet: RTMP Audio, size: 96
 D/RtmpConnection: wrote packet: RTMP Audio, size: 96

```

**Comment by thegobot** [comment]
After updating the YASEA, problems became more ... I think that there are any features of the protocol...
This problem is only with the adobe flash player enabled buffered
I also checked the rtmpdump - normally

Example:
[Flash Player with buffering](https://www.youtube.com/watch?v=W1x-vW3jnXY&feature=youtu.be)
[Flash Player without buffering](https://www.youtube.com/watch?v=hlGUZy8CD1U)

**Comment by begeekmyfriend** [comment]
Sorry I have made a mistake on video frame cache counting due to the preceding RTMP writer thread optimization. Please sync https://github.com/begeekmyfriend/yasea/commit/da16e33f2ddaa17ebb30a296c93dd22d751f2ab0

**Comment by begeekmyfriend** [comment]
And another measure is reducing the output bitrate to relax player buffer https://github.com/begeekmyfriend/yasea/commit/3c341b6a0ea2957ef5f5756cce0f77e90e3d7804

**Comment by begeekmyfriend** [comment]
Note Yasea has been public for several months. Seldom guys have the same issue like yours. So I think you had better check up your networking.

**Comment by thegobot** [comment]
Thanks!
Now I updated and check
Again, the problem is only enabled buffering.  In adobe flash(action script 3) has standard class NetStream
```
ns = new  NetStream()
ns.bufferTime = 1;
ns.bufferTimeMax = 3;
```
Only with these SETUP, in one case YASEA works poorly. When cached frames fall sharply in the network and when the stream was stopped and restarted. The problem is specific to Flash, but it is the most popular solution

Another question. What advise to analyze packets rtmp coming to the client? I tried Wireshark, but it is poorly analyzed rtmp

**Comment by begeekmyfriend** [comment]
Choose the right RTMP server (like Nginx-RTMP) with Wireshark. Some server will produce AV packets which will be shown as UNKNOWN packets through Wireshark.

Another way is to reuse and modify RTMP module of Yasea in which the handler threading will parse the packets from reading stream and log.

**Comment by thegobot** [comment]
Thanks! I updated yasya...
Excellent! Now the frames are in the cache after NetStream.Publish.Start!

I noticed some bugs ...
1) With enabled needToFindKeyFrame = true the frameCache ALWAIS EMPTY
2) Ok... i disabled needToFindKeyFrame = false... Now only sent AUDIO. Video frames are not added rtmp_write_packet()

```
 // when sps or pps not sent, ignore the packet.
// @see https://github.com/simple-rtmp-server/srs/issues/203

if (!h264_sps_pps_sent) {
   Log.e(TAG, "BAD");
    return;
}
```
logs...
```
 D/RtmpConnection: wrote packet: RTMP Audio, size: 94
 D/RtmpConnection: wrote packet: RTMP Audio, size: 89
 D/RtmpConnection: wrote packet: RTMP Audio, size: 99
 E/SrsFlvMuxer: BAD
 D/RtmpConnection: wrote packet: RTMP Audio, size: 96
 D/RtmpConnection: wrote packet: RTMP Audio, size: 91
 E/SrsFlvMuxer: BAD
 D/RtmpConnection: wrote packet: RTMP Audio, size: 110
 E/SrsFlvMuxer: BAD
 D/RtmpConnection: wrote packet: RTMP Audio, size: 93
 D/RtmpConnection: wrote packet: RTMP Audio, size: 91
 D/RtmpConnection: wrote packet: RTMP Audio, size: 96
 D/RtmpConnection: wrote packet: RTMP Audio, size: 91
 D/RtmpConnection: wrote packet: RTMP Audio, size: 91
 D/RtmpConnection: wrote packet: RTMP Audio, size: 92
 D/RtmpConnection: wrote packet: RTMP Audio, size: 92
 E/SrsFlvMuxer: BAD
 E/SrsFlvMuxer: BAD
 D/RtmpConnection: wrote packet: RTMP Audio, size: 94
 D/RtmpConnection: wrote packet: RTMP Audio, size: 92
 D/RtmpConnection: wrote packet: RTMP Audio, size: 123
 D/RtmpConnection: wrote packet: RTMP Audio, size: 96
 D/RtmpConnection: wrote packet: RTMP Audio, size: 91
 E/SrsFlvMuxer: BAD
 D/RtmpConnection: wrote packet: RTMP Audio, size: 90
 D/RtmpConnection: wrote packet: RTMP Audio, size: 93
 D/RtmpConnection: wrote packet: RTMP Audio, size: 92
 E/SrsFlvMuxer: BAD
 D/RtmpConnection: wrote packet: RTMP Audio, size: 118
 D/RtmpConnection: wrote packet: RTMP Audio, size: 86
```

**Comment by begeekmyfriend** [comment]
You had better learn some basic knowledge of the format of H.264 and you will know what the sps and pps mean to the video stream.

**Comment by thegobot** [comment]
Yes, but the previous version YASEA worked and the video was sent... Not now.
h264_pps and h264_sps always null
```
if (h264_pps == null || h264_sps == null) {                
    return;
}
```

**Comment by begeekmyfriend** [comment]
Fixed https://github.com/begeekmyfriend/yasea/commit/59931aabf5b390d8c2745ef8501d2304b6bb50de. As you can see, we still cannot prevent sps and pps within the first frame produced by the encoder being added into the cache even before the connection finished. The connection and encoder operation are not in the same thread. You can control the encoder start and close outsider by yourself.

**Comment by thegobot** [comment]
Yes thank you. It works now!

I understand what the problem is with the sps and pps. Also, I have proposed a [solution ](https://github.com/begeekmyfriend/yasea/pull/111#issuecomment-246103408) for me and maybe for someone
Just run the encoder after a full connection. This requires a new event in [RtmpConnection](https://github.com/begeekmyfriend/yasea/blob/master/app/src/main/java/net/ossrs/yasea/rtmp/io/RtmpConnection.java#L566)

Also, I have compiled the [SRS](https://github.com/ossrs/srs/tree/2.0release) and test it

**Comment by thegobot** [comment]
Yes, the problem exists only with nginx-rtmp. SRS vs Flash has no such problem
